File: R/landings/smooth_sspn_ratio.cpp
// A multivariate state-space random walk model to smooth observed ratios of
// shortspine thornyhead to total identified thornyhead. The predicted smoothed
// ratios are estimated as random effects. Observation error is assumed to
// inversely proportional to the unidentified catch (larger catches are assigned
// smaller CVs). An inverse logit transformation is applied to the ratios in order
// bound random effects between 0 and 1. The only fixed effects parameters in
// the model are the standard deviation of the process error.

#include <TMB.hpp>
#include <iostream>

template<class Type>
Type objective_function<Type>::operator() ()
{
  // model dimensions
  DATA_IVECTOR(model_yrs);

  // observed ratios of sspn
  DATA_MATRIX(logit_ratio_obs);
  DATA_MATRIX(ratio_cv);

  // indexing vector
  DATA_IVECTOR(pointer_PE); // length = ncol ratio obs (# fleets), unique values = number of PE parameters

  // parameter section

  // PARAMETER(dummy);  // dummy var for troubleshooting
  PARAMETER_VECTOR(log_PE); // process errors

  // random effects of predicted sspn ratios
  PARAMETER_MATRIX(logit_ratio_pred);

  // negative log likelihood
  vector<Type> jnll(2); // random walk, observation error
  jnll.setZero();
  Type nll = 0;

  // derived quantities 
  int nyrs = model_yrs.size();
  int nfleets = logit_ratio_obs.cols();

  matrix<Type> ratio_pred(nyrs, nfleets);
  ratio_pred.setZero();
  for(int i = 0; i < nyrs; i++) {
    for(int j = 0; j < nfleets; j++) {
      ratio_pred(i,j) = invlogit(logit_ratio_pred(i,j));
    }
  }

  matrix<Type> ratio_obs(nyrs, nfleets);
  ratio_obs.setZero();
  for(int i = 0; i < nyrs; i++) {
    for(int j = 0; j < nfleets; j++) {
      ratio_obs(i,j) = invlogit(logit_ratio_obs(i,j));
    }
  }

  matrix<Type> log_ratio_sd(nyrs, nfleets);
  log_ratio_sd.setZero();

  for(int i = 0; i < nyrs; i++) {
    for(int j = 0; j < nfleets; j++) {
      log_ratio_sd(i,j) = ratio_cv(i,j) * ratio_cv(i,j) + Type(1.0);
      log_ratio_sd(i,j) = sqrt(log(log_ratio_sd(i,j)));
    }
  }

  // random effects contribution to likelihood
  for(int i = 1; i < nyrs; i++) {
    for(int j = 0; j < nfleets; j++) {
      jnll(0) -= dnorm(log(ratio_pred(i-1,j)), log(ratio_pred(i,j)), exp(log_PE(pointer_PE(j))), 1);
    }
  }

  // likelihood for sspn ratio data observation error
  for(int i = 0; i < nyrs; i++) {
    for(int j = 0; j < nfleets; j++) {

      if(ratio_obs(i,j) > 0) {
        jnll(1) -= dnorm(logit_ratio_obs(i,j), logit_ratio_pred(i,j), log_ratio_sd(i,j), 1);
      }

    }
  }

  // report section
  ADREPORT(ratio_pred);
  REPORT(jnll);
  // jnll = dummy * dummy;        // Uncomment when debugging code
  nll = jnll.sum();
  return nll;

}

File: R/landings/smooth_sspn_ratio.R
# Code to run the smooth_sspn_ratio.cpp program, a multivariate state-space
# random walk model used to smooth the time series of SSPN ratios
# Contact: jane.sullivan@noaa.gov
# Last updated March 2023

library(tidyverse)
library(TMB)
library(scales)

scale_fill_colorblind7 = function(.ColorList = 2L:8L, ...){
  scale_fill_discrete(..., type = colorblind_pal()(8)[.ColorList])
}

scale_color_colorblind7 = function(.ColorList = 2L:8L, ...){
  scale_color_discrete(..., type = colorblind_pal()(8)[.ColorList])
}

# data ----

id_thorny <- read_csv('outputs/fishery_data/identified_catch.csv')
prop_sspn <- read_csv('outputs/fishery_data/prop_sspn.csv') %>% 
  dplyr::filter(!state_gear %in% c('WA_Trawl', 'WA_Non-trawl')) 

# the CVs of the sspn ratio data are assumed to be inversely proportional to the
# size of the identified catch of thornyheads (sspn + lspn). we assume that the
# smallest catch within each fleet has a CV=0.3 and the largest catch has a
# CV=0.05
id_thorny <- id_thorny %>% 
  tidyr::complete(species, year, state_gear, fill = list(mtons = 0)) %>%
  dplyr::group_by(year, fleet = state_gear) %>% 
  dplyr::summarise(catch = sum(mtons)) %>% 
  dplyr::group_by(fleet) %>% 
  dplyr::mutate(cv = scales::rescale(catch, to = c(0.3, 0.05))) 

id_thorny %>% pivot_wider(id_cols = year, names_from = fleet, values_from = cv)
         
# new
prop_dat <- prop_sspn %>% 
  dplyr::select(year, fleet = state_gear, prop_sspn) %>% 
  dplyr::left_join(id_thorny) %>% 
  dplyr::select(-catch) %>% 
  dplyr::mutate(cv = ifelse(is.na(cv), 0.3, cv))

# data list for TMB ----
dat <- list(model_yrs = sort(unique(prop_dat$year)))

tmp <- prop_dat %>% 
  tidyr::pivot_wider(id_cols = year, names_from = fleet, values_from = prop_sspn, values_fill = NA) %>% # print(n=Inf)
  dplyr::select(-year) %>% 
  as.matrix()
tmp[tmp == 1] <- 0.9999; tmp[tmp == 0] <- 0.0001
dat$logit_ratio_obs <- log((tmp)/(1-tmp)) # logit transformation

dat$ratio_cv <- prop_dat %>% 
  tidyr::pivot_wider(id_cols = year, names_from = fleet, values_from = cv, values_fill = NA) %>% 
  dplyr::select(-year) %>% 
  as.matrix()

dat$pointer_PE <- 0:(ncol(dat$ratio_cv)-1) # unique process error estimated for each fleet. tmb/c++ indexing starts at 0
# dat$pointer_PE <- rep(0, ncol(dat$ratio_cv)) # only process error estimated (simplest model)

# parameter list for TMB ----
par <- list()

# process error starting value = 1
par$log_PE <- rep(log(1), length(unique(dat$pointer_PE)))

# predicted sspn ratios (estimated as random effects in logit space). use
# interpolated values as starting values. note that values cannot exactly equal
# 1 or 0 using this approach
tmp <- apply(X = tmp, MARGIN = 2, FUN = zoo::na.approx, maxgap = 100, rule = 2)
tmp[tmp == 1] <- 0.9999; tmp[tmp == 0] <- 0.0001
par$logit_ratio_pred <- log((tmp)/(1-tmp)) # logit transformation

# map list for TMB ----

map <- par
map$log_PE <- as.factor(1:length(map$log_PE))

map$logit_ratio_pred <- as.factor(1:length(map$logit_ratio_pred))

# id parameters that should be estimated as random effects
random = 'logit_ratio_pred'

# fit model ----

# need to run these if you don't have the program compiled
# setwd('R/landings')
# TMB::compile('smooth_sspn_ratio.cpp')
# dyn.load(TMB::dynlib('smooth_sspn_ratio'))

obj <- TMB::MakeADFun(data = dat, parameters = par, random = random,
                     DLL = 'smooth_sspn_ratio', map = map)  

# lower and upper bounds relate to obj$par and must be the same length as obj$par
opt <- nlminb(start = obj$par, objective = obj$fn, hessian = obj$gr)

sdrep <- TMB::sdreport(obj)
summary(sdrep)

# plot results ----
out <- data.frame(name = rownames(summary(sdrep)),
                  estimate = data.frame(summary(sdrep))$Estimate,
                  se = data.frame(summary(sdrep))$Std..Error)

pred <- tidyr::expand_grid(fleet = colnames(dat$logit_ratio_obs),
                   variable = 'pred_prop_sspn',
                   year = dat$model_yrs) %>% 
  dplyr::left_join(prop_dat) %>% 
  tidyr::separate(col = fleet, sep = '_', into = c('state', 'gear'), remove = FALSE) %>%
  dplyr::mutate(ratio_pred = out$estimate[out$name == 'ratio_pred'],
                logit_ratio_pred = out$estimate[out$name == 'logit_ratio_pred'],
                sd_logit_ratio_pred = out$se[out$name == 'logit_ratio_pred'],
                pred_uci = 1 / (1 + exp(-logit_ratio_pred - 1.96 * sd_logit_ratio_pred)),
                pred_lci = 1 / (1 + exp(-logit_ratio_pred + 1.96 * sd_logit_ratio_pred)))
              
ggplot(pred, aes(x = year, col = gear, fill = gear)) +
  geom_point(aes(y = prop_sspn)) +
  geom_ribbon(aes(ymin = pred_lci, ymax = pred_uci), col = NA, alpha = 0.2) +
  geom_line(aes(y = ratio_pred)) +
  scale_color_colorblind7() +
  scale_fill_colorblind7() +
  facet_wrap(~state, ncol = 1) +
  labs(x = 'Year', y = 'Proportion', title = 'Porportion shortspine thornyhead in identified catch',
       fill = 'Gear', col = 'Gear')

pred %>% write_csv('outputs/fishery_data/smooth_sspn_ratios.csv')
ggsave("outputs/fishery_data/smooth_sspn_ratios.png", dpi=300, height=8, width=10, units='in')
